enhancements/storage/vsphere-storage-check.md

```diff
@@ -0,0 +1,228 @@
+### Non-Goals
+
+* Any automatic fix of the environment.
+
+## Proposal
+
+Run a dedicated vsphere-monitoring-operator that periodically tests OCP / vSphere:
+
+* The operator reports metrics, some of them sent via telemetry to us.
+* The operator reports alerts for common configuration issues.
+* The operator logs performed checks + more detailed info about OCP/vSphere config, so it's available in must-gather.
+
+Add pre-flight checks to installer.
+
+### Checks
+
+These checks will be performed regularly and / or in the installer:
+
+| Test | Periodic | Installer |
+|-|-|-|
+| OCP can list recent tasks. | yes | yes |
+| OCP can list folders in the default datastore. |  yes | yes |
+| Kubelets run with cloud-provider configured (i.e. have `Node.Spec.ProviderID`). |  yes | no |
+| All VMs have `disk.enableUUID=TRUE`. |  yes | no |
+| Datastore names used in OCP config, StorageClasses and PVs are not too long. |  yes | yes |
+| Stretch goal: OCP can create a volume + attach it to a VM (this usually works when the checks above pass). |  yes | no |
+| Stretch goal: Was a VM / storage recently migrated using vMotion? We don't support this (yet). | yes | no |
+| Stretch goal: Is Storage Distributed Resource Scheduler enabled? We don't support this. | yes | no |
+
+When a previous test failed, the initial test frequency is once per 1 hour, doubling with every failed attempt up to once per 8 hours max.
+When a previous test succeeded, the frequency is once per 8 hours.
+
+### Metrics
+
+We want these metrics to be transferred via telemetry to us:
+
+* List of failed periodic tests.
+* HW version of vSphere VMs.
+* vCenter version.
+* ESXi host version(s).
+
+### User Stories
+
+#### Story 1
+
+As a support person, I can see in must-gather that OCP is configured correctly on vSphere.
+
+I.e. results of the check(s) should be logged in must-gather.
+
+#### Story 2
+
+As OCP cluster admin, I want to know if I can create PVs and use them in pods. I.e. get an alert when something gets broken.
+
+#### Story 3
+
+As OCP PM, I want to know how many customers use an old vSphere version that may not be supported by a future OCP update and decide if I should push engineering harder to support the old vSphere version or convince few customers to upgrade.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+The operator will have direct access to vSphere API, which does not fine-grained permissions. As result, the permissions may be quite high. On the other hand, the operator exposes very little surface to attack.
+
+## Design Details
+
+All these are subject of change, see Open Questions below.
+
+* It is a standalone operator.
+* Running in namespace `openshift-cluster-storage-operator` (TBD).
+* Started by cluster-storage-operator (CSO) when it detects it runs on vSphere  (TBD).
+* It reports its own health to CSO via CRD named `VMWareStorageMonitoring`. Provisional API:
+    ```go
+    type VMWareStorageMonitoring struct {
```

Comment:
should we call it `VMWareStorageMonitoring`? I would prefer if we kept it as `vSphereMonitoring` may be.. because at the end of the day we are monitoring more than just storage..

Reply:
Sure, changed to `VSphereMonitoring`.